pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java

```diff
@@ -224,19 +141,126 @@ public void readValues(int[] rows, int rowStartPos, int rowSize, int[] values, i
      * at a time, identifying the rows in sequence that belong to the same block. This logic is more complex, but may
      * perform better in the sorted case.
      *
-     * An alternative is to not have multiple buffers, but just copy the values from one buffer to next as we
+     * An alternative is to not have multiple _dataBuffers, but just copy the values from one buffer to next as we
      * increase the number of rows.
      */
-    if (readers.size() == 1) {
-      readers.get(0).getReader().readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
+    if (_readers.size() == 1) {
+      _readers.get(0).getReader().readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
     } else {
-      for (int rowIter = rowStartPos, valueIter = valuesStartPos;
-          rowIter < rowStartPos + rowSize;
+      for (int rowIter = rowStartPos, valueIter = valuesStartPos; rowIter < rowStartPos + rowSize;
           rowIter++, valueIter++) {
         int row = rows[rowIter];
         int bufferId = getBufferId(row);
-        values[valueIter] = readers.get(bufferId).getInt(row);
+        values[valueIter] = _readers.get(bufferId).getInt(row);
+      }
+    }
+  }
+
+  private int getBufferId(int row) {
+    return row / _numRowsPerChunk;
+  }
+
+  private void addBuffer() {
+    LOGGER.info("Allocating {} bytes for column {} for segment {}", _chunkSizeInBytes, _columnName,
+        _memoryManager.getSegmentName());
+    PinotDataBuffer buffer = _memoryManager.allocate(_chunkSizeInBytes, _columnName);
+    _capacity += buffer.size();
+
+    buffer.order(ByteOrder.nativeOrder());
+    _dataBuffers.add(buffer);
+
+    FixedByteSingleValueMultiColReader reader =
+        new FixedByteSingleValueMultiColReader(buffer, _numRowsPerChunk, new int[]{_columnSizesInBytes});
+    FixedByteSingleValueMultiColWriter writer =
+        new FixedByteSingleValueMultiColWriter(buffer, _numRowsPerChunk, /*cols=*/1, new int[]{_columnSizesInBytes});
+
+    final int startRowId = _numRowsPerChunk * (_dataBuffers.size() - 1);
+    _writers.add(new WriterWithOffset(writer, startRowId));
+    _readers.add(new ReaderWithOffset(reader, startRowId));
+  }
+
+  private void addBufferIfNeeded(int row) {
```

Comment:
why is this computation needed for every row? when we add a new buffer, we can just compute the max number of rows right and check for that?

Reply:
Yes, good point. Will replace _capacity with _capacityInRows, so we can first check the row number against that, and then do the more expensive computation if required.